File: C++/checkPossibility.cpp
/**
 * @File    : checkPossibility.cpp
 * @Brief   : 非递减数列 贪心策略各种情况下是否依然是最优解
 * @Link    : https://leetcode-cn.com/problems/non-decreasing-array/
 * @Author  : Wei Li
 * @Date    : 2021-07-06
*/

#include <iostream>
#include <vector>


class Solution
{
public:
    bool checkPossibility(std::vector<int>& nums)
    {
        int cnt = 0;
        for (size_t i = 0; i < nums.size() - 1; ++i)
        {
            int x = nums[i], y = nums[i+1];
            if (x > y)
            {
                cnt++;
                if (cnt > 1)
                {
                    return false;
                }
                if (i > 0 && y < nums[i-1])
                {
                    nums[i + 1] = x;
                }
            }
        }
        return true;        
    }
};


// ------------------------------
int main(int argc, char** argv)
{
    // std::vector<int> vec_nums = {4, 2, 3};
    std::vector<int> vec_nums = {4, 2, 1};

    auto solution = Solution();
    bool check_ok = solution.checkPossibility(vec_nums);
    std::cout << "The solution of this problem is : " << std::boolalpha << check_ok << std::endl;
    
    return 0;
}

File: Python/checkPossibility.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: checkPossibility.py
@Function: 非递减数列 贪心策略各种情况下是否依然是最优解
@Link: https://leetcode-cn.com/problems/non-decreasing-array/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-06
"""

class Solution:
    def checkPossibility(self, nums):
        cnt = 0
        for i in range(len(nums)-1):
            x, y = nums[i], nums[i+1]
            if x > y:
                cnt += 1
                if cnt > 1:
                    return False
                if i > 0 and y < nums[i-1]:
                    nums[i+1] = x
        return True


# ---------------------------------
if __name__ == "__main__":
    # nums = [4,2,3]
    nums = [4,2,1]

    solution = Solution()
    check = solution.checkPossibility(nums)
    print(f"The solution of this problem is : {check}")